### ISSUE TITLE
Add TLS support for checkServerIdentity

### ISSUE THREAD

**Issue by TazmanianD** [brianc/node-postgres#1185]
This is partly a question as well as a feature request. We are using a hosted Postgres service and they seemed to have made the annoying decision to self-sign their certificates and not put the proper hostnames in the certificate (the hostname is for the actual database which is behind a "portal" proxy). The result is that if we try to use the public certificate they provide us, Node rejects it because of the hostname mismatch.

Apparently Node does provide a mechanism to work around this through the use of a "checkServerIdentity" function. However, because this pg module whitelists the properties it passes along to the Node TLS module, we can't make use of it.

The following code is an example:

```
const fs = require('fs');
const pg = require('pg');

const config = {
  user: 'xxx',
  database: 'xxx',
  password: 'xxx',
  host: 'xxx',
  port: xxx,

  ssl: {
    rejectUnauthorized: true,
    ca: fs.readFileSync('my-certificate.ca'),
  },
};

const client = new pg.Client(config);

client.connect((error) => {
  if (error) {
    console.error(error);
  } else {
    console.log('Connected!');
  }
});
```

For us, this generates the following error: `Error: Hostname/IP doesn't match certificate's altnames: "Host: xxx. is not in the cert's altnames: DNS:yyy`.

This shouldn't be a security issue for us because we're using a pinned certificate and as long as the server presents a certificate that matches, it shouldn't be an issue that it's not configured with the correct hostname.

The following proposal (if passed through to the TLS sub-system) would fix this problem:

```
  ssl: {
    rejectUnauthorized: true,
    ca: fs.readFileSync('my-certificate.ca'),

    checkServerIdentity: () => {
    },
  },
```

The issue seems to lie in https://github.com/brianc/node-postgres/blob/master/lib/connection.js#L91 where's it's copying specific ssl properties. Is there a reason this doesn't just copy the entire ssl object instead of whitelisting known properties?

Is this an acceptable solution to this problem and are there any workarounds without this support?

**Comment by TazmanianD** [comment]
I have found a workaround that can be used until this feature is implemented. You can patch the Node TLS library:

```
const tls = require('tls');
const originalCheckServerIdentity = tls.checkServerIdentity;

tls.checkServerIdentity = function checkServerIdentity(host, cert) {
  // Only ignore the host check if it's one of our expected hosts
  const matchesServer = host.match(/<regexp that matches your servers>/) !== null;
  if (matchesServer) {
    // Returning undefined means the check is good
    return undefined;
  } else {
    return originalCheckServerIdentity(host, cert);
  }
};
```

**Comment by charmander** [comment]
Made available (with a bug when not provided) in pg 7.4.2 by #1444.
